addons/binding/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/handler/HeosThingBaseHandler.java

```diff
@@ -0,0 +1,275 @@
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.heos.handler;
+
+import static org.openhab.binding.heos.HeosBindingConstants.*;
+import static org.openhab.binding.heos.internal.resources.HeosConstants.*;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Map;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.PercentType;
+import org.eclipse.smarthome.core.library.types.PlayPauseType;
+import org.eclipse.smarthome.core.library.types.RawType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.io.net.http.HttpUtil;
+import org.openhab.binding.heos.internal.HeosChannelHandlerFactory;
+import org.openhab.binding.heos.internal.api.HeosFacade;
+import org.openhab.binding.heos.internal.api.HeosSystem;
+import org.openhab.binding.heos.internal.handler.HeosChannelHandler;
+import org.openhab.binding.heos.internal.resources.HeosEventListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link HeosThingBaseHandler} class is the base Class all HEOS handler have to extend.
+ * It provides basic command handling and common needed methods.
+ *
+ * @author Johannes Einig - Initial contribution
+ *
+ */
+public abstract class HeosThingBaseHandler extends BaseThingHandler implements HeosEventListener {
+
+    protected String id;
+    protected HeosSystem heos;
+    protected HeosFacade api;
+    protected HeosChannelHandlerFactory channelHandlerFactory;
+    protected HeosBridgeHandler bridge;
+
+    private long refreshStartTime;
+    private long refreshRequestTime;
+    private static final int REFRESH_BLOCK_TIME = 5000;
+
+    private Logger logger = LoggerFactory.getLogger(HeosThingBaseHandler.class);
+
+    public HeosThingBaseHandler(Thing thing, HeosSystem heos, HeosFacade api) {
+        super(thing);
+        this.heos = heos;
+        this.api = api;
+        setId();
+    }
+
+    /**
+     * To be implemented by extending class by the command
+     * for updating the HEOS thing via the {@link HeosSystem}. As example
+     * for a player the command would be {@link HeosSystem#getPlayerState(String)}
+     */
+    protected abstract void updateHeosThingState();
+
+    /**
+     * The channels which has to be updated during the refresh command
```

Comment:
Note that a REFRESH command is sent for a specific channel and you should ONLY update that channel. And it can happen that you get many REFRESH commands for different channels sent subsequently.